// react-admin-firebase/dist/src/providers/lazy-loading/paramsToQuery.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { limit, orderBy, query, startAfter, where, } from 'firebase/firestore';
import { findLastQueryCursor, getQueryCursor } from './queryCursors';
var defaultParamsToQueryOptions = {
    filters: true,
    sort: true,
    pagination: true,
};
export function paramsToQuery(collection, params, resourceName, flogger, options) {
    if (options === void 0) { options = defaultParamsToQueryOptions; }
    return __awaiter(this, void 0, void 0, function () {
        var filterConstraints, sortConstraints, paginationConstraints, _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    filterConstraints = options.filters
                        ? getFiltersConstraints(params.filter)
                        : [];
                    sortConstraints = options.sort ? getSortConstraints(params.sort) : [];
                    if (!options.pagination) return [3 /*break*/, 2];
                    return [4 /*yield*/, getPaginationConstraints(collection, __spreadArray(__spreadArray([], filterConstraints, true), sortConstraints, true), params, resourceName, flogger)];
                case 1:
                    _a = _b.sent();
                    return [3 /*break*/, 3];
                case 2:
                    _a = [];
                    _b.label = 3;
                case 3:
                    paginationConstraints = _a;
                    return [2 /*return*/, {
                            noPagination: query.apply(void 0, __spreadArray([collection], __spreadArray(__spreadArray([], filterConstraints, true), sortConstraints, true), false)),
                            withPagination: query.apply(void 0, __spreadArray([collection], __spreadArray(__spreadArray(__spreadArray([], filterConstraints, true), sortConstraints, true), paginationConstraints, true), false)),
                        }];
            }
        });
    });
}
export function getFiltersConstraints(filters) {
    return Object.entries(filters).flatMap(function (_a) {
        var fieldName = _a[0], fieldValue = _a[1];
        if (Array.isArray(fieldValue)) {
            return [where(fieldName, 'array-contains-any', fieldValue)];
        }
        else if (Object.keys(filters).length === 1 && isNaN(fieldValue)) {
            return [
                where(fieldName, '>=', fieldValue),
                where(fieldName, '<', fieldValue + 'z'),
            ];
        }
        else {
            return [where(fieldName, '==', fieldValue)];
        }
    });
}
export function getSortConstraints(sort) {
    if (sort != null && sort.field !== 'id') {
        var field = sort.field, order = sort.order;
        var parsedOrder = order.toLocaleLowerCase();
        return [orderBy(field, parsedOrder)];
    }
    return [];
}
function getPaginationConstraints(collectionRef, queryConstraints, params, resourceName, flogger) {
    return __awaiter(this, void 0, void 0, function () {
        var _a, page, perPage, queryCursor;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    _a = params.pagination, page = _a.page, perPage = _a.perPage;
                    if (!(page === 1)) return [3 /*break*/, 1];
                    return [2 /*return*/, [limit(perPage)]];
                case 1: return [4 /*yield*/, getQueryCursor(collectionRef, params, resourceName, flogger)];
                case 2:
                    queryCursor = _b.sent();
                    if (!!queryCursor) return [3 /*break*/, 4];
                    return [4 /*yield*/, findLastQueryCursor(collectionRef, queryConstraints, params, resourceName, flogger)];
                case 3:
                    queryCursor = _b.sent();
                    _b.label = 4;
                case 4: return [2 /*return*/, [startAfter(queryCursor), limit(perPage)]];
            }
        });
    });
}
export function getFullParamsForQuery(reactAdminParams, softdeleteEnabled) {
    return __assign(__assign({}, reactAdminParams), { filter: softdeleteEnabled
            ? __assign({ deleted: false }, reactAdminParams.filter) : reactAdminParams.filter });
}
export function getNextPageParams(params) {
    return __assign(__assign({}, params), { pagination: __assign(__assign({}, params.pagination), { page: params.pagination.page + 1 }) });
}
